### ISSUE TITLE
Coverage for dynamically-generated code

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [istanbuljs/nyc#371]
So, I've got a bit of an unusual situation: I have _generated_ code, that I want to cover. (And to boot, it's all being passed through Babel.)

This normally wouldn't make sense; but my code is ‘generated’ only by exclusion: a file is read in (and passed through Mustache of all things … don't ask), and then blocks of code are contextually removed from it. This is all exhaustively tested in a generative fashion (i.e. that all API entrypoints function as expected, no matter which combinations of chunks are removed from the generative code.)

Specifically, the code is [read in and manipulated](https://github.com/ELLIOTTCABLE/giraphe/blob/ce377320/giraphe.es6.js#L68-L93), then [eventually](https://github.com/pierrec/node-eval/blob/25c75035778910a781a5486bb4d396d3df526121/eval.js#L57-L66) passed through `vm.runInNewContext()` by [the `'eval'` package](https://github.com/pierrec/node-eval):

``` es6
const body = Mustache.render(walkTemplate, options)
    , desc = options._description.join('+')
    , path = walkTemplatePath + (desc ? '.'+desc : '')

const wrap = require_eval(body, path, {}, true).default

const func = wrap(symbols, options)
```

For probably-obvious reasons, I want to ensure that all the _actual_ lines of code (lines in the source-file) are eventually tested. I'm trying to do this with Istanbul / `nyc` / `babel-plugin-istanbul`; but I'm running into trouble.

---

I have Babel configured as suggested, i.e. invoking the `istanbul` babel-plugin only when in `NODE_ENV=coverage` (which my test-script is designed to configure),

```
 , "babel": {
      ...
    , "sourceMaps"             : "inline"
    , "compact"                : false

    , "env": {
         "coverage": {
            "plugins": [ ..., "istanbul" ]
         }
         ...
      }
   }
```

… and I have `nyc` configured to include two specific files (I'm not sure if I'm doing this right? The documentation is … sparse, to say the least):

```
 , "nyc": {
      "all"                    : true
    , "include"                : ["giraphe.es6.js", "walk.es6.js.mustache"]
    , "extension"              : [".es6.js", ".es6.js.mustache"]

    , "require"                : ["babel-register"]

    , "sourceMap"              : false
    , "instrument"             : false
   }
```

The convoluted setup is working for the _root_ file … but I'm getting nothing for the `walk.es6.js.mustache` file, despite the fact that Babel's recognizing and transforming it just fine, and the fact that I explicitly mention it in the `"include"`s.
### **tl;dr**: What's my next step, to get Istanbul to pick up and record coverage for a source-file that's evaluated with `vm.runInNewContext`?

---
#### Forensics:

Mac OS X, **`10.11.6`** “El Capitan”
Node.js **`v6.4.0`**
npm **`3.10.3`**

`npm ls`: https://gist.github.com/ELLIOTTCABLE/1dff1e4c06d75e7b5360ec4a2fdbc24f

> ([Here's](https://github.com/ELLIOTTCABLE/giraphe/blob/Current+/walk.es6.js.mustache) the source-file for the code being generated; and [here's](https://gist.github.com/ELLIOTTCABLE/79c2dbbcd8eb3f302ef92b6586d2f2f3) my work-so-far on the test-script that generates coverage. If it looks over-complicated, that's because I'm [copying](https://github.com/ELLIOTTCABLE/Paws.js/blob/Current+/Scripts/test.sh) most of the effort from another, much more complicated, project.)

**Comment by bcoe** [comment]
Hey @ELLIOTTCABLE have been away for a few days on vacation, but will start digging into this for you soon; if no one beats me to the punch.

**Comment by ELLIOTTCABLE** [comment]
@bcoe Never apologize for having a life outside of programming.

This is non-critical for me; even this much attention is impressive. Take your time. <3

**Comment by JaKXz** [comment]
@coreyfarrell any thoughts on this one?

**Comment by coreyfarrell** [comment]
Sorry I missed this one.  The initial description mentioned `vm.runInNewContext`.  I suspect that installing `@istanbuljs/nyc-config-hook-run-in-this-context` from npm and configuring nyc to use that as base config could solve this issue:
```json
{
  "extends": "@istanbuljs/nyc-config-hook-run-in-this-context"
}
```

Worth noting: nyc depends on having a unique filename to associate coverage with, and that filename must be permitted by the include/exclude/extension settings.  I'm unsure what would happen if the same filename is used multiple times for different chunks of code, you could get odd merges or it could even crash the reporters.

Running `node-eval` without providing a filename is likely an unsupportable scenario for nyc, as is using the same filename multiple times for different chunks of code

**Comment by maritz** [comment]
@coreyfarrell Sadly @istanbuljs/nyc-config-hook-run-in-this-context didn't fix it for me.

I tried giving different or random filenames as well, just to see what happens. Nothing changed.

I might very well be doing something fundamentally wrong, but I lack the understanding and time to try and debug it currently. :-(
